{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skbio import DNA\n",
    "from skbio.sequence.distance import kmer_distance\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#Checking Distance similarity of sequence, i.e. finding match, mismatch or gaps\n",
    "\n",
    "#Hamming Distance - must have same length\n",
    "\n",
    "#Checking for dissimilarity\n",
    "#Used in sequence clustering, phylogenetic reconstruction\n",
    "#Mismatch and match frequency\n",
    "#Kmer-distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n",
      "[ True  True  True False False  True  True  True  True]\n",
      "2\n",
      "[False False False  True  True False False False False]\n",
      "0.2222222222222222\n",
      "2\n",
      "{'ACT': 1, 'CTA': 1, 'TAT': 1, 'ATC': 1, 'TCA': 1, 'CAG': 1, 'AGT': 1}\n",
      "{'ACT': 1, 'CTT': 1, 'TTA': 1, 'TAC': 1, 'ACA': 1, 'CAG': 1, 'AGT': 1}\n",
      "\n",
      "Scikit Bio Kmer distance method Result: 0.7272727272727273\n",
      "Custom Kmer distance function 1 Result: 0.7272727272727273\n",
      "Custom Kmer distance function 2 Result: 0.7272727272727273\n",
      "Use Symmetric Difference in Set for Kmer Distance: 0.7272727272727273\n"
     ]
    }
   ],
   "source": [
    "seq1 = DNA(\"ACTATCAGT\")\n",
    "seq2 = DNA(\"ACTTACAGT\")\n",
    "\n",
    "#Calculate matches,mismatches frequency\n",
    "print(seq1.match_frequency(seq2))\n",
    "\n",
    "#Get actual matches, returns comparisions between each letter as Bool\n",
    "print(seq1.matches(seq2))\n",
    "\n",
    "#Get frequency for mismatches\n",
    "print(seq1.mismatch_frequency(seq2))\n",
    "\n",
    "#Get actual mismatches, returns comparisions between each letter as Bool\n",
    "print(seq1.mismatches(seq2))\n",
    "\n",
    "#Get Hamming Distance using Scikit Bio - explanation in Biopython program\n",
    "print(seq1.distance(seq2))\n",
    "\n",
    "\n",
    "#Custom Hamming distance Function\n",
    "def hamming_distance(lhs,rhs):\n",
    "    return len([(x,y) for x,y in zip(lhs,rhs) if x != y])\n",
    "\n",
    "print(hamming_distance(seq1,seq2))\n",
    "\n",
    "\n",
    "#Kmer distance between 2 sequences is the fraction of kmers that are unique to either sequence.\n",
    "#Find Kmers not found in each sequence: the sum of non-intersection\n",
    "#Using set ^ set for dissimilar/symmetric_difference\n",
    "\n",
    "#Finds and counts number of unique slices of sequence given, argument specifies length of unique slice to find i.e. k\n",
    "#Can be used to compare similarity between each sequence\n",
    "seq1_freq = seq1.kmer_frequencies(3)\n",
    "seq2_freq = seq2.kmer_frequencies(3)\n",
    "print(seq1_freq)\n",
    "print(seq2_freq)\n",
    "\n",
    "#kmer distance\n",
    "print(\"\\nScikit Bio Kmer distance method Result:\",kmer_distance(seq1,seq2,k=3))\n",
    "\n",
    "#Kmer distance custom function: Set, intersect and dissimilar set\n",
    "\n",
    "#Method 1\n",
    "def custom_kmer_distance1(seq1,seq2):\n",
    "    seq1_set, seq2_set = set(seq1), set(seq2)\n",
    "    #dissimilarity seq = union seq - intersection seq\n",
    "    dissimilarity = len(seq1_set.union(seq2_set)) - len(seq1_set.intersection(seq2_set))\n",
    "    unique_distance = dissimilarity/len(seq1_set.union(seq2_set))\n",
    "    return unique_distance\n",
    "\n",
    "\n",
    "#Method 2\n",
    "def custom_kmer_distance2(seq1,seq2):\n",
    "    seq1_set, seq2_set = set(seq1), set(seq2)\n",
    "    #dissimilarity seq = union seq - intersection seq\n",
    "    #dissimilarity = len(seq1_set.union(seq2_set)) - len(seq1_set.intersection(seq2_set))\n",
    "    dissimilarity = seq1_set ^ seq2_set\n",
    "    unique_distance = len(dissimilarity)/len(seq1_set.union(seq2_set))\n",
    "    return unique_distance\n",
    "\n",
    "#Use the frequency dictionaries to calculate the distance\n",
    "print(\"Custom Kmer distance function 1 Result:\",custom_kmer_distance1(seq1_freq,seq2_freq))\n",
    "print(\"Custom Kmer distance function 2 Result:\",custom_kmer_distance2(seq1_freq,seq2_freq))\n",
    "\n",
    "#Symmetric difference in set\n",
    "print(\"Use Symmetric Difference in Set for Kmer Distance:\",len(set(seq1_freq).symmetric_difference(set(seq2_freq)))/len(set(seq1_freq).union(set(seq2_freq))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'C': 0.2222222222222222, 'A': 0.3333333333333333}\n",
      "{'A': 3, 'C': 2, 'T': 3, 'G': 1}\n",
      "Custom function for finding kmers in seq 1: {'A': 3, 'C': 2, 'T': 3, 'G': 1}\n",
      "Custom function for finding kmers in seq 2: {'ACT': 1, 'CTT': 1, 'TTA': 1, 'TAC': 1, 'ACA': 1, 'CAG': 1, 'AGT': 1}\n",
      "\n",
      "Iterating through kmers\n",
      "ACT\n",
      "CTA\n",
      "TAT\n",
      "ATC\n",
      "TCA\n",
      "CAG\n",
      "AGT\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAD4CAYAAADFAawfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAM30lEQVR4nO3dYaxk9VnH8e+vbBGxtZpyxRS4XDSAqVQtubYKWipoxS6lEDFlTRtR66bGIjaauiYmtNHEfdGYkkiMG6u2aS1pKKjpNrQ1BbURaVnEUHZbirAiIGmbvqhSDTY+vphZOt3ee2fuzBz22fL9JBPunHP+zzzn7J9fzj1zZm6qCklSX8851g1IkrZmUEtScwa1JDVnUEtScwa1JDW3Y4iip5xySq2trQ1RWpK+KR04cOCLVbWy0bpBgnptbY277757iNKS9E0pyb9tts5LH5LUnEEtSc0Z1JLUnEEtSc0Z1JLUnEEtSc0Z1JLUnEEtSc0Z1JLU3CCfTNT2rO3Zv3CNw3t3LqETSR15Ri1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1Jzc0U1EnekuT+JJ9O8v4kJw3dmCRpZGpQJzkN+HVgvarOA04Arh66MUnSyKyXPnYA35pkB3Ay8PhwLUmSJu2YtkFVPZbkHcAjwH8DH62qjx69XZLdwG6A1dXVZfc5t7U9+xeucXjvziV0IknzmeXSx3cCrwXOAl4EfFuS1x+9XVXtq6r1qlpfWVlZfqeS9Cw1y6WPnwQerqovVNX/ArcAFwzbliTpiFmC+hHgR5KcnCTAJcChYduSJB0xNair6i7gZuAe4L7xmH0D9yVJGpv6ZiJAVV0PXD9wL5KkDfjJRElqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqbqagTvIdSW5O8pkkh5L86NCNSZJGdsy43Q3AbVV1VZITgZMH7EmSNGFqUCf5duAVwDUAVfUU8NSwbUmSjpjljPp7gC8Af57kB4EDwHVV9eTkRkl2A7sBVldXl91nK2t79i80/vDenUvqRMejZc+fRettVFO9zHKNegdwPvDHVfVS4Elgz9EbVdW+qlqvqvWVlZUltylJz16zBPWjwKNVddf4+c2MgluS9AyYGtRV9QTw70nOHS+6BDg4aFeSpKfNetfHtcD7xnd8PAT84nAtSZImzRTUVXUvsD5wL5KkDfjJRElqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqzqCWpOYMaklqbtY/xSUdF9b27F9o/OG9O5dab6Oa0nZ5Ri1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1Jzc0c1ElOSPLPST40ZEOSpK+3nTPq64BDQzUiSdrYTEGd5HRgJ/Cnw7YjSTrajhm3eyfwVuD5m22QZDewG2B1dXXuhtb27J97LMDhvTsXGq9nzqL/1uC/91D8/7CXqWfUSS4DPl9VB7barqr2VdV6Va2vrKwsrUFJerab5dLHhcDlSQ4DNwEXJ3nvoF1Jkp42Nair6neq6vSqWgOuBj5eVa8fvDNJEuB91JLU3qxvJgJQVXcAdwzSiSRpQ55RS1JzBrUkNWdQS1JzBrUkNWdQS1JzBrUkNWdQS1JzBrUkNWdQS1JzBrUkNWdQS1JzBrUkNWdQS1JzBrUkNWdQS1JzBrUkNWdQS1Jz2/oLLzp+rO3Zv3CNw3t3LqETSYvyjFqSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmjOoJak5g1qSmpsa1EnOSHJ7kkNJ7k9y3TPRmCRpZJY/bvtV4Der6p4kzwcOJPlYVR0cuDdJEjOcUVfVf1TVPeOf/xM4BJw2dGOSpJFZzqiflmQNeClw1wbrdgO7AVZXV5fQmrpZ27N/ofGH9+5cUifS4vMRvnFOdp3jM7+ZmOR5wAeB36iqLx+9vqr2VdV6Va2vrKwss0dJelabKaiTPJdRSL+vqm4ZtiVJ0qRZ7voI8C7gUFX94fAtSZImzXJGfSHwBuDiJPeOH68euC9J0tjUNxOr6hNAnoFeJEkb8JOJktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktScQS1JzRnUktTc1L/wIknLsLZn/0LjD+/duaROjj+eUUtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtScwa1JDVnUEtSczMFdZJLk3w2yYNJ9gzdlCTpa6YGdZITgBuBnwFeDOxK8uKhG5MkjcxyRv0y4MGqeqiqngJuAl47bFuSpCNSVVtvkFwFXFpVbxw/fwPw8qp681Hb7QZ2j5+eC3x2+e0CcArwxeY17bFnveOlpj32rDdUzSPOrKqVjVbsmGFwNlj2DeleVfuAfdtsbNuS3F1V651r2mPPesdLTXvsWW+omrOY5dLHo8AZE89PBx4fph1J0tFmCepPAWcnOSvJicDVwN8M25Yk6Yiplz6q6qtJ3gx8BDgB+LOqun/wzjY3xOWVZde0x571jpea9tiz3lA1p5r6ZqIk6djyk4mS1JxBLUnNtQ/qJFcmqSTfl+QlSe4dP76U5OHxz387Q50XTox9IsljSe4bPz84sezINifOWfPp8ZO9z7vPRy1/S5L/SfKCReslOTvJh5L8a5IDSW5P8ooZ652a5C+TPDQee2eSK8frXpbkjiSfS3JPkv1JXjJnzddNHM//Gn+Nwb1J3rPAfp+T5MPjr0M4lOQDSU6dsd53J7lpfMwOjuucs+Satyf5yhLn+JHnqxv1vsh+j9dte05ucRwHmZPj9TeMj8csn8be6jhu+TqDqarWD+ADwD8Abztq+V8AV81Z823Ab01btoSaG/a+wD5/crz8mkXqAScBDwCXT2xz3ix1Gd1XfyfwpollZwLXAqcCh4ELJtb9GHDFvDUnnt8BrC9hvz8HvGZim58Azptzv38I+PEhai5zjk97nQV73NacnHIclz4nxz8/B3gE+CfglUs+jl83T4d6DFp84ebgecBjwDnAZ45at5RJvNWyRWpu1fs8+wx8L3AQuAj4yCL1gF8G3j3nfl4C/N0m634PePsya05scwfbCOpN9vuXgPfMud8XA3+/wfKl15xYv5Q5Pu115u1xnjm5xXEcZE5OrP8w8AvAnyxwHKfO06Ee3S99XAHcVlUPAF9Kcv6xbmgb5u19s3G7gPczOns5N8l3LVDv+4F7Zt2Ro2w1dt66i/SzmY32+zzgwJz1Nhs7RM1lG6rHeebkZvWGmpPwtT5vBS5L8tyBXmcw3YN6F6MvgWL8313HsJftmrf3zcZdDdxUVf8H3AL83LL6SHJrkk8nuWXGmpNjb0zyL0k+tcG6u8bXbG9YVs1tOJ7nzvFi3jk51bLmZEbvNb0a+Kuq+jJwF/CqJfW4jHk6k1m+6+OYSPJCRr8mnZekGH3YppK8tca/h3Q1b+9bjHsvcDbwsSQAJwIPMfr62W33AbwdePpNmqq6Msk68I4Zdu9+4Gcnxv5aklOAu4HbgPOBvx6ve3lGX+p12QI1t23Kfl80T81xj1dtsnzZNZdtkdfZcGySH2COOblFL/czzJy8FHgBcN+4z5OBrwD7Z6i7ndcZVOcz6qsYXfs7s6rWquoM4GFGb051N2/vm417J6M3xNbGjxcBpyU5c856DwAXJrl8YtuTZ9y3jwMnJfnVDcbeCFyT5IJt1t2q5jy22u8Lkuw8smFGfxRj6l0p4x6/JcmvTIz9YeDBZddMMm/wD/E6m+33Dcw3J7c6jkPMyV3AG4/0CZwFvCrJPPNr2fN0Zp2Deheja0qTPgj8/DHoZbvm7X2zcWsbLL+V0a+e8/ZxGfCm8W1GdwK/C/z+lHqMfyO4Argoo1vHPgm8G/jtqnoCeB3wBxndqvaPjELzj+atOa2fTUzb72szun3wIHAN8PlpBcc9Xgn81Pj2sfsZvdH0+EA1l2aR19li7CuZY07OcByXOSevB36aibPnqnoS+ATwmml1t/E6887TmfkRcklqrvMZtSQJg1qS2jOoJak5g1qSmjOoJak5g1qSmjOoJam5/wfk/B9Ub7gpFQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Using Kmers and Kmers frequency\n",
    "#L will have L-k+1 kmers and n^k total possible kmers - useful for custom function\n",
    "#Useful for assembling DNA sequences\n",
    "#improve heterogeneous gene expression\n",
    "#identify species in metagenomic samples\n",
    "#Create attenuated vaccines\n",
    "\n",
    "#Seq 1 = ATGA...\n",
    "#1st Kmer of 3 is ATG, next Kmer of 3 is TGA\n",
    "\n",
    "#Check the frequency for specific nucleotides\n",
    "print(seq1.frequencies(chars={\"A\",\"C\"}, relative=True))\n",
    "\n",
    "print(seq1.kmer_frequencies(k=1))\n",
    "\n",
    "#Custom function to find kmers\n",
    "def count_kmers(seq, k=3):\n",
    "    counts, number_kmers = {}, len(seq)-k+1\n",
    "    for num in range(number_kmers):\n",
    "        #slicing sequence together\n",
    "        kmer = str(seq)[num:num+k]\n",
    "        if kmer not in counts:\n",
    "            counts[kmer] = 0\n",
    "        counts[kmer] += 1\n",
    "    return counts\n",
    "\n",
    "print(\"Custom function for finding kmers in seq 1:\",count_kmers(seq1,1))\n",
    "print(\"Custom function for finding kmers in seq 2:\",count_kmers(seq2))\n",
    "\n",
    "#Iterating through kmers\n",
    "print(\"\\nIterating through kmers\")\n",
    "for k in seq1.iter_kmers(3, overlap=True):\n",
    "    print(k)\n",
    "\n",
    "#Plot kmers frequency\n",
    "dna_seq1 = DNA(\"ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTTGTAGATCTGTTCTCTAAA\")\n",
    "plt.bar(dna_seq1.kmer_frequencies(2).keys(), dna_seq1.kmer_frequencies(2).values())\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
